app/controllers/lti_controller.rb

```diff
@@ -25,31 +26,125 @@ def get_canvas_config
+      login_hint: params[:login_hint],
+      response_mode: 'form_post',
+      nonce: nonce,
+      prompt: 'none',
+      state: session.id
+    }
+    referrer = URI(request.referer)
+
+    # TODO: generalize this to platforms other than canvas.
+    auth_request_uri = URI("#{referrer.scheme}://#{referrer.host}:#{referrer.port}/api/lti/authorize_redirect")
+
+    http = Net::HTTP.new(auth_request_uri.host, auth_request_uri.port)
+    req = Net::HTTP::Post.new(auth_request_uri)
+    req.set_form_data(auth_params)
+
+    res = http.request(req)
+    location = res['location']
+    redirect_to location
+  end
+
+  def redirect_login
+    referrer_uri = URI(request.referer)
+    known_jwks = %w[q.utoronto.ca canvas.instructure.com] # only supporting canvas right now
+    if params[:id_token].blank? || params[:state] != session.id.to_s || known_jwks.exclude?(referrer_uri.host)
+      render 'shared/http_status', locals: { code: '422', message: I18n.t('lti.config_error') }, layout: false
+      return
+    end
+
+    # Get canvas JWK set
+    jwk_url = "#{referrer_uri.scheme}://#{referrer_uri.host}:#{referrer_uri.port}/api/lti/security/jwks"
+    canvas_jwks = JSON.parse(Net::HTTP.get_response(URI(jwk_url)).body)
+    begin
+      decoded_token = JWT.decode(
+        params[:id_token],
+        nil,
+        true, # Verify the signature of this token
+        algorithms: ['RS256'],
+        iss: "#{referrer_uri.scheme}://#{referrer_uri.host}",
+        verify_iss: true,
+        aud: session[:client_id], # canvas uses client ID as the aud parameter
+        verify_aud: true,
+        jwks: canvas_jwks
+      )
+    rescue JWT::DecodeError
+      render 'shared/http_status', locals: { code: '422', message: I18n.t('lti.config_error') }, layout: false
+      return
+    end
+    session[:lti_course_id] = decoded_token[0]['https://purl.imsglobal.org/spec/lti/claim/custom']['course_id']
+    session[:lti_user_id] = decoded_token[0]['https://purl.imsglobal.org/spec/lti/claim/custom']['user_id']
+    session[:lti_course_name] = decoded_token[0]['https://purl.imsglobal.org/spec/lti/claim/custom']['course_name']
+    deployment_id = decoded_token[0]['https://purl.imsglobal.org/spec/lti/claim/deployment_id']
+    lti_host = URI(request.referer).host
+    lti_client = Lti.find_or_create_by(client_id: session[:client_id], deployment_id: deployment_id, host: lti_host)
+    session[:lti_client_id] = lti_client.id
+    redirect_to root_path
   end
 
   def public_jwk
     # TODO: implement this function
     render json: { state: 'success' }
   end
 
+  def choose_course
+    if request.post?
+      begin
+        course = Course.find(params[:course])
+        lti = Lti.find(session[:lti_client_id])
+        lti.update!(course: course)
+      rescue StandardError
+        flash_message(:error, 'Unsuccessful. Please relaunch MarkUs from your LMS.')
+        render 'choose_course'
+      else
+        flash_message(:success, 'Successfully linked courses')
```

Comment:
I feel like this message can be improved

Reply:
I updated this message to "Success. `%(markus_course_name)` is now linked with your LMS". I still feel like it could be improved, but I couldn't come up with better wording.